### PR TITLE
Add '#nullable enable' to PublicAPI.(Un)shipped.txt

### PR DESCRIPTION
**Pull request by Youssef1313**

### CODE REVIEW

**Comment by jcouv** [comment]
I already have a PR/draft that does this, but it's blocked because of an issue with source generators which only affects the C# compiler (not the common compiler code). So I think it's fine to move ahead with this PR for the common compiler code first.
The diagnostics with oblivious APIs need to be mitigated with pragmas for now. See https://github.com/dotnet/roslyn/pull/41000

File: src/Compilers/Core/Portable/.editorconfig [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[*.cs]
+dotnet_diagnostic.RS0041.severity = none
```
Comment (Youssef1313) [inline src/Compilers/Core/Portable/.editorconfig:2]: Do you want a tracking issue to remove this?

Comment (jcouv) [reply src/Compilers/Core/Portable/.editorconfig:2]: We don't need to add a blanket suppression. A few files need a pragma. See [https://github.com/dotnet/roslyn/pull/41000](https://github.com/dotnet/roslyn/pull/41000)

---
In reply to: [565227031](https://github.com/dotnet/roslyn/pull/50793#discussion_r565227031) [](ancestors = 565227031)

File: src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs [hunk 2]
```diff
@@ -566,7 +566,9 @@ private SyntaxNodeOrToken[] NodesWithSeparators
             get { return _list.ToArray(); }
         }
 
+#pragma warning disable RS0041 // uses oblivious reference types
         public Enumerator GetEnumerator()
+#pragma warning restore RS0041 // uses oblivious reference types
```
Comment (Youssef1313) [inline src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs:571]: I'm confused by the RS0041 produced here. There is no `#nullable disable` in this file, and NRT should be enabled. Not sure why there is a diagnostic here.

Comment (jcouv) [reply src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs:571]: `Enumerator` is a nested type. It's usage here is treated like:
```
#nullable disable
SeparatedSyntaxList<T>
#nullable enable 
    .Enumerator
```

See https://github.com/dotnet/roslyn-analyzers/pull/4681 for some illustrations.
This only happens if `T` is is constrained to a reference type if I recall correctly.

This is a problem with the symbol API which we likely won't be able to fix.

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 3]
```diff
@@ -1,24 +1,25 @@
-Microsoft.CodeAnalysis.GeneratorAttribute.GeneratorAttribute(string firstLanguage, params string[] additionalLanguages) -> void
-Microsoft.CodeAnalysis.GeneratorAttribute.Languages.get -> string[]
-Microsoft.CodeAnalysis.GeneratorExecutionContext.SyntaxContextReceiver.get -> Microsoft.CodeAnalysis.ISyntaxContextReceiver
-Microsoft.CodeAnalysis.GeneratorInitializationContext.RegisterForSyntaxNotifications(Microsoft.CodeAnalysis.SyntaxContextReceiverCreator receiverCreator) -> void
+#nullable enable
```
Comment (jcouv) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:1]: nit: technically `#nullable enable` is only needed in one of the files. I'd put it just in the shipped file.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 8)

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review. This is enabling tracking of nullability in shipped files for common compiler project.

Note that a couple of APIs have oblivious. They are tracked with leading `~` in the API text file and require a suppression in code.

Here are the APIs that still involve oblivious in this project (copy/pasted from search in this PR):
```
~Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader.AddDependencyLocation(string fullPath) -> void
~Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader.LoadFromPath(string fullPath) -> System.Reflection.Assembly
~Microsoft.CodeAnalysis.ISymbol.ContainingAssembly.get -> Microsoft.CodeAnalysis.IAssemblySymbol
~Microsoft.CodeAnalysis.ISymbol.ContainingModule.get -> Microsoft.CodeAnalysis.IModuleSymbol
~Microsoft.CodeAnalysis.ISymbol.ContainingNamespace.get -> Microsoft.CodeAnalysis.INamespaceSymbol
~Microsoft.CodeAnalysis.ISymbol.ContainingSymbol.get -> Microsoft.CodeAnalysis.ISymbol
~Microsoft.CodeAnalysis.ISymbol.ContainingType.get -> Microsoft.CodeAnalysis.INamedTypeSymbol
~Microsoft.CodeAnalysis.SeparatedSyntaxList<TNode>.GetEnumerator() -> Microsoft.CodeAnalysis.SeparatedSyntaxList<TNode>.Enumerator
~Microsoft.CodeAnalysis.SyntaxList<TNode>.GetEnumerator() -> Microsoft.CodeAnalysis.SyntaxList<TNode>.Enumerator
```

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review of community PR. Thanks

**Review by jaredpar — APPROVED** [review]

**Comment by jcouv** [comment]
@Youssef1313 The correctness leg still spots a problem.
There is a known issue with source generators and analyzers, which cause the publicAPI analyzer/fixer not to annotate all the APIs. Maybe that's what you're hitting. The only solution at this point is to manually annotate the handful of APIs that  need it.
I think you can repro those correctness build errors locally with `build.cmd -runAnalyzers`.

**Comment by Youssef1313** [comment]
> Maybe that's what you're hitting.

@jcouv Nope, these are new APIs added after my PR (i.e. they exist on master but not on my branch).

I'll merge master into my branch and run the fix all again.

**Comment by jcouv** [comment]
Squashed/merged. Thanks for the contribution @Youssef1313

**Pull request merged**